src/firmware/linux/guest/types.rs

```diff
@@ -1 +1,388 @@
+    pub key: [u8; 32],
+}
+
+/// Information provided by the guest owner for requesting an attestation
+/// report and associated certificate chain from the AMD Secure Processor.
+#[repr(C)]
+pub struct SnpExtReportReq {
+    /// Address of the [`SnpReportReq`].
+    pub data: u64,
+
+    /// Address of extended guest request certificate chain.
+    pub certs_address: u64,
+
+    /// Length of the buffer which will hold the certificates.
+    pub certs_len: u32,
+}
+
+#[derive(Default)]
+#[repr(C)]
+pub struct SnpExtReportRsp {
+    /// Address of the [`SnpReportRsp`].
+    pub data: u64,
+
+    /// Address of certificate chain.
+    pub certs_address: u64,
+
+    /// Length of the buffer which will hold the certificates.
+    pub certs_len: u32,
+}
+
+/// Information provided by the guest owner for requesting an attestation
+/// report from the AMD Secure Processor.
+#[repr(C, packed)]
+pub struct SnpReportReq {
+    /// Guest-provided data to be included int the attestation report
+    pub report_data: [u8; 64],
+
+    /// The VMPL to put into the attestation report. Must be greater than or
+    /// equal to the current VMPL and at most three.
+    pub vmpl: u32,
+
+    /// Reserved memory slot, must be zero.
+    reserved: [u8; 28],
+}
+
+impl SnpReportReq {
+    /// Instantiates a new [`SnpReportReq`] for fetching an [`AttestationReport`] from the PSP.
+    ///
+    /// # Arguments
+    ///
+    /// * `report_data` - (Optional) 64 bytes of unique data to be included in the generated report.
+    /// * `vmpl` - The VMPL level the guest VM is running on.
+    pub fn new(report_data: Option<[u8; 64]>, vmpl: u32) -> Self {
+        Self {
+            report_data: report_data.unwrap_or([0; 64]),
+            vmpl,
+            reserved: Default::default(),
+        }
+    }
+}
+
+/// The response from the PSP containing the generated attestation report.
+/// According to the kernel, this should match [`self::guest::ReportRsp`].
+/// The Report is padded to exactly 4000 Bytes to make sure the page size
+/// matches.
+///
+///
+/// ```txt
+///     96 Bytes (*Message Header)
+/// + 4000 Bytes (*Encrypted Message)
+/// ------------
+///   4096 Bytes (4K Memory Page Alignment)
+/// ```
+/// <sup>*[Message Header - 8.26 SNP_GUEST_REQUEST - Table 97](<https://www.amd.com/system/files/TechDocs/56860.pdf#page=113>)</sup>
+///
+/// <sup>*[Encrypted Message - sev-guest.h](<https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/sev-guest.h>)</sup>
+#[repr(C)]
+pub struct SnpReportRsp {
+    /// The status of key derivation operation.
+    ///     0h: Success.
+    ///     16h: Invalid parameters.
+    pub status: u32,
+    /// Size in bytes of the report.
+    pub report_size: u32,
+    reserved_0: [u8; 24],
+    /// The attestation report generated by the firmware.
+    pub report: AttestationReport,
+    /// Padding bits to meet the memory page alignment.
+    reserved_1: [u8; 2780],
+}
```

Comment:
If we have something like `static_assert!` (the proposed rust additions to the kernel have it), then we can add:

```
static_assert!(core::mem::size_of::<SnpReportRsp>() == 4000);
```

Reply:
Yes, I like that! Will add in next commit.